### PR TITLE
My attempt to complete #367 (NKRO)

### PR DESCRIPTION
**Pull request by benedekkupper**
This branch is a rebase-refactor-review of the #367 original, and the subsequent `nkey-rollover` branch. The commits are restructured in a way that at each commit the keyboard should work at least as well as before. I still don't have the means to test this branch, so please do that before merging. In case of problems, please bisect the commit that breaks something.

### CODE REVIEW

**Comment by kareltucek** [comment]
@benedekkupper uhk pretty much stops working with this branch. 

Bisecting leads to:

```
$ git bisect bad
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[3e244f9555dd64505c867bbec43c54ba90650e00] remove external report scancode indexing
$ git bisect good
1b53f367307803932b74dd8269932f5fd115ed57 is the first bad commit
commit 1b53f367307803932b74dd8269932f5fd115ed57
Author: Benedek Kupper <kupper.benedek@gmail.com>
Date:   Thu Dec 2 23:13:08 2021 +0100

    hid report descriptors: put them in .rodata

    The report descriptors are fixed, so mark them const in order to stop wasting
    precious RAM just for storing a copy of them.

    Signed-off-by: Benedek Kupper <kupper.benedek@gmail.com>

 .../src/usb_descriptors/usb_descriptor_basic_keyboard_report.c |  2 +-
 .../src/usb_descriptors/usb_descriptor_basic_keyboard_report.h |  2 +-
 right/src/usb_descriptors/usb_descriptor_generic_hid_report.c  |  2 +-
 right/src/usb_descriptors/usb_descriptor_generic_hid_report.h  |  2 +-
 right/src/usb_descriptors/usb_descriptor_hid.c                 | 10 +++++-----
 .../src/usb_descriptors/usb_descriptor_media_keyboard_report.c |  2 +-
 .../src/usb_descriptors/usb_descriptor_media_keyboard_report.h |  2 +-
 right/src/usb_descriptors/usb_descriptor_mouse_report.c        |  2 +-
 right/src/usb_descriptors/usb_descriptor_mouse_report.h        |  2 +-
 .../usb_descriptors/usb_descriptor_system_keyboard_report.c    |  2 +-
 .../usb_descriptors/usb_descriptor_system_keyboard_report.h    |  2 +-
 11 files changed, 15 insertions(+), 15 deletions(-)
```

**Edit: removed some misleading parts**

**Comment by kareltucek** [comment]
With benedekkuper/nkro (unmodified), I am getting:
<details>
<summary>dmesg</summary>

```
[36768.355351] usb 1-4: new full-speed USB device number 82 using xhci_hcd
[36768.509560] usb 1-4: New USB device found, idVendor=1d50, idProduct=6124, bcdDevice= 1.02
[36768.509575] usb 1-4: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[36768.509581] usb 1-4: Product: UHK 60 v2
[36768.509585] usb 1-4: Manufacturer: Ultimate Gadget Laboratories
[36768.513338] hid (null): global environment stack underflow
[36768.513546] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513567] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513583] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513598] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513613] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513629] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513644] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513659] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513674] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513690] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513705] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513720] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513735] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513750] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513766] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513781] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513797] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513812] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513831] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513850] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513870] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x0
[36768.513889] hid-generic 0003:1D50:6124.00E0: unknown main item tag 0x3
[36768.513908] hid-generic 0003:1D50:6124.00E0: global environment stack underflow
[36768.513915] hid-generic 0003:1D50:6124.00E0: item 0 0 1 11 parsing failed
[36768.513955] hid-generic: probe of 0003:1D50:6124.00E0 failed with error -22
[36768.515130] hid (null): global environment stack underflow
[36768.515621] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515646] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515664] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515683] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515703] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515735] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515742] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515749] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515756] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515763] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515769] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515777] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515784] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515790] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515797] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515804] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515811] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515818] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515824] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515830] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515837] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515844] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515850] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515858] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515864] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x6
[36768.515871] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x5
[36768.515878] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515885] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515892] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515898] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515906] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x0
[36768.515912] hid-generic 0003:1D50:6124.00E1: unknown main item tag 0x3
[36768.515919] hid-generic 0003:1D50:6124.00E1: global environment stack underflow
[36768.515922] hid-generic 0003:1D50:6124.00E1: item 0 0 1 11 parsing failed
[36768.515936] hid-generic: probe of 0003:1D50:6124.00E1 failed with error -22
[36768.516212] hid (null): global environment stack underflow
[36768.516262] hid-generic 0003:1D50:6124.00E2: collection stack underflow
[36768.516269] hid-generic 0003:1D50:6124.00E2: item 0 4 0 12 parsing failed
[36768.516279] hid-generic: probe of 0003:1D50:6124.00E2 failed with error -22
[36768.516643] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516650] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516656] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516661] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516667] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516672] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516678] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516683] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516689] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516694] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516700] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516705] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516711] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516717] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516724] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516729] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516740] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516745] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516751] hid-generic 0003:1D50:6124.00E3: unknown main item tag 0x0
[36768.516825] hid-generic 0003:1D50:6124.00E3: hidraw2: USB HID v1.10 Device [Ultimate Gadget Laboratories UHK 60 v2] on usb-0000:00:14.0-4/input3
[36768.517203] hid (null): unknown global tag 0xd
[36768.517331] hid (null): global environment stack underflow
[36768.517357] hid (null): global environment stack underflow
[36768.517403] hid (null): global environment stack underflow
[36768.517486] hid-generic 0003:1D50:6124.00E4: unknown main item tag 0x0
[36768.517497] hid-generic 0003:1D50:6124.00E4: unknown main item tag 0x0
[36768.517504] hid-generic 0003:1D50:6124.00E4: unknown main item tag 0x0
[36768.517512] hid-generic 0003:1D50:6124.00E4: unknown main item tag 0x0
[36768.517520] hid-generic 0003:1D50:6124.00E4: unknown global tag 0xd
[36768.517523] hid-generic 0003:1D50:6124.00E4: item 0 0 1 13 parsing failed
[36768.517539] hid-generic: probe of 0003:1D50:6124.00E4 failed with error -22
```
</details>

**Comment by kareltucek** [comment]
With proper patch, at (1b53f367307803932b74dd8269932f5fd115ed57) :


<details>
<summary>patch</summary>

```
index d3874fa4..ff283f1c 100644
--- a/right/src/usb_interfaces/usb_interface_basic_keyboard.c
+++ b/right/src/usb_interfaces/usb_interface_basic_keyboard.c
@@ -238,8 +238,8 @@ void UsbBasicKeyboard_MergeReports(const usb_basic_keyboard_report_t* sourceRepo
 void UsbBasicKeyboard_ForeachScancode(const usb_basic_keyboard_report_t* report, void(*action)(uint8_t))
 {
     if (UsbBasicKeyboardGetProtocol() == USB_HID_BOOT_PROTOCOL) {
-        for (uint8_t i = 0; (i < ARRAY_SIZE(report->boot.scancodes)) && (report->boot.scancodes[i] != 0); i++) {
-            action(report->boot.scancodes[i]);
+        for (uint8_t i = 0; (i < ARRAY_SIZE(report->scancodes)) && (report->scancodes[i] != 0); i++) {
+            action(report->scancodes[i]);
         }
     }
 }
```
</details>

<details>
<summary>dmesg</summary>

```
[37887.372883] usb 1-4: new full-speed USB device number 116 using xhci_hcd
[37887.523113] usb 1-4: New USB device found, idVendor=1d50, idProduct=6124, bcdDevice= 1.02
[37887.523130] usb 1-4: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[37887.523137] usb 1-4: Product: UHK 60 v2
[37887.523142] usb 1-4: Manufacturer: Ultimate Gadget Laboratories
[37887.529246] hid (null): unknown global tag 0xd
[37887.529726] hid-generic 0003:1D50:6124.0146: unknown main item tag 0x0
[37887.529746] hid-generic 0003:1D50:6124.0146: unknown global tag 0xd
[37887.529753] hid-generic 0003:1D50:6124.0146: item 0 0 1 13 parsing failed
[37887.529780] hid-generic: probe of 0003:1D50:6124.0146 failed with error -22
[37887.530433] hid (null): unknown global tag 0xd
[37887.531061] hid-generic 0003:1D50:6124.0147: unknown main item tag 0x0
[37887.531080] hid-generic 0003:1D50:6124.0147: unknown global tag 0xd
[37887.531086] hid-generic 0003:1D50:6124.0147: item 0 0 1 13 parsing failed
[37887.531112] hid-generic: probe of 0003:1D50:6124.0147 failed with error -22
[37887.531896] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x2
[37887.531917] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.531933] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.531948] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.531962] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.531980] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.531995] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532011] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532026] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532041] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532057] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532072] hid-generic 0003:1D50:6124.0148: unknown main item tag 0x0
[37887.532294] hid-generic 0003:1D50:6124.0148: hidraw2: USB HID v1.10 Device [Ultimate Gadget Laboratories UHK 60 v2] on usb-0000:00:14.0-4/input2
[37887.542128] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542155] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542179] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542199] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542220] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542240] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542264] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542285] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542305] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542324] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542344] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542365] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542385] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542404] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542424] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542444] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542465] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542484] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542504] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542524] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542544] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542564] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542584] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542604] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542624] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542643] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542663] hid-generic 0003:1D50:6124.0149: unknown main item tag 0x0
[37887.542922] hid-generic 0003:1D50:6124.0149: hidraw3: USB HID v1.10 Device [Ultimate Gadget Laboratories UHK 60 v2] on usb-0000:00:14.0-4/input3
[37887.544014] hid-generic 0003:1D50:6124.014A: collection stack underflow
[37887.544040] hid-generic 0003:1D50:6124.014A: item 0 4 0 12 parsing failed
[37887.544067] hid-generic: probe of 0003:1D50:6124.014A failed with error -22

```   
</details>

**Comment by benedekkupper** [comment]
@kareltucek thanks for the testing, I didn't expect it that this part would fail. The branch is now rebased, the main change is that the `USB_DESC_STORAGE_TYPE` define in `usb_api.h` controls what type the descriptors use. I also now changed all descriptors to be in .rodata, and set forced 4-byte alignment for all descriptors, that's the only possible problem I can imagine with this change. If it isn't, then you will see even bigger errors during USB enumeration, in this case please just test the rest with `USB_DESC_STORAGE_TYPE` set to `uint8_t`.

**Comment by kareltucek** [comment]
@benedekkupper const version yields following in dmesg:

```
[58573.048158] usb 1-4: new full-speed USB device number 78 using xhci_hcd
[58573.176312] usb 1-4: device descriptor read/64, error 18
[58573.412219] usb 1-4: device descriptor read/64, error 18
[58573.648240] usb 1-4: new full-speed USB device number 79 using xhci_hcd
[58573.776317] usb 1-4: device descriptor read/64, error 18
[58574.016164] usb 1-4: device descriptor read/64, error 18
[58574.124366] usb usb1-port4: attempt power cycle
[58574.776076] usb 1-4: new full-speed USB device number 80 using xhci_hcd
[58574.796370] usb 1-4: device descriptor read/8, error -61
[58574.924882] usb 1-4: device descriptor read/8, error -61
[58575.164047] usb 1-4: new full-speed USB device number 81 using xhci_hcd
[58575.184441] usb 1-4: device descriptor read/8, error -61
[58575.312953] usb 1-4: device descriptor read/8, error -61
[58575.420334] usb usb1-port4: unable to enumerate USB device
```

With `uint8_t` everything seems to work fine (just throws tons of `warning: 'Usb*KeyboardReportDescriptor' defined but not used` during compilation).

**Comment by benedekkupper** [comment]
FFS who designs such pieces of silicon/SDK? As evidenced by [this](https://community.nxp.com/t5/Kinetis-Microcontrollers/Freescale-USB-stack-v4-1-1/m-p/328342) and [this post](https://community.nxp.com/t5/Kinetis-Microcontrollers/USB-Stack-Putting-enumeration-strings-in-FLASH/m-p/311103), the USB controller doesn't have access to flash by default, but this can be changed with a flag.

**Comment by benedekkupper** [comment]
Can you test with the new commit on top? Also, "everything seems to work fine" means that there is also no missed first key press, as in your branch?

**Comment by mondalaci** [comment]
@benedekkupper Why can't you test the firmware?

**Comment by benedekkupper** [comment]
I still haven't been able to set up the agent's libraries so the makefile can run the necessary scripts from it for flashing. Note that I can flash from the agent itself, but the agent UI requires a packaged firmware, which the makefile itself doesn't produce.

**Comment by kareltucek** [comment]
> Can you test with the new commit on top? 

Works and compiles, no warnings!

> Also, "everything seems to work fine" means that there is also no missed first key press, as in your branch?

Yes, no missed scancode.

File: right/src/usb_interfaces/usb_interface_basic_keyboard.c [hunk 1]
```diff
@@ -136,3 +145,146 @@ usb_status_t UsbBasicKeyboardCallback(class_handle_t handle, uint32_t event, voi
...
+    } else {
+        for (uint8_t i = 0; i < ARRAY_SIZE(targetReport->bitfield); i++) {
+            targetReport->bitfield[i] |= sourceReport->bitfield[i];
+        }
+    }
+}
+
+void UsbBasicKeyboard_ForeachScancode(const usb_basic_keyboard_report_t* report, void(*action)(uint8_t))
+{
+    // TODO: do we need to consider modifiers?
```
Comment (kareltucek) [inline right/src/usb_interfaces/usb_interface_basic_keyboard.c:276]: (No. This is fine.)

**Comment by kareltucek** [comment]
Also... great job with the refactors!

**Comment by benedekkupper** [comment]
I imagine that there would still be errors with macros if the keyboard's protocol is changed at runtime, but in reality there should be few cases where the boot protocol is activated, so I guess we can deal with all the corner cases later.

**Comment by mondalaci** [comment]
I've just tested this PR, and NKRO is working flawlessly on Linux, Mac, Windows 10, and in the BIOS of my Ryzen PC. Excellent job!

Upon merging this PR, I can close #367, right?

**Comment by benedekkupper** [comment]
Yes, this ticket supersedes #367 (all the changes are incorporated from there), and closes #70.

**Pull request merged**